File: packages/flutter-heddle/lib/src/hot_plane.dart
// lib/src/hot_plane.dart — WeftHotPlane: the consumer side of HPL1 in Dart.
//
// Zero-allocation reads: readLane fills a CALLER-OWNED WeftLaneSnapshot;
// readRecent fills a caller Float64List. Tears are counted (never swallowed,
// Law 4). Every multi-byte access passes Endian.little explicitly (Law 2).
//
// Attach paths:
//   * WeftHotPlane.fromBytes(Uint8List)      — Dart-side copy / test rigs
//   * WeftHotPlane.fromPointer(ptr, bytes)   — dart:ffi native memory
//     (Engineer 1's SharedArrayBuffer-equivalent bridge; POSIX shm mmap)

import 'dart:typed_data';
import 'hpl1_layout.dart';

/// Caller-owned lane snapshot (allocate once, reuse forever — Law 1).
class WeftLaneSnapshot {
  int seqLo = 0, seqHi = 0;
  double current = 0, min = 0, max = 0, avg = 0;
  int samplesSeenLo = 0, samplesSeenHi = 0;
  int head = 0, flags = 0;
  int publishNsLo = 0, publishNsHi = 0;
  int drops = 0;
  int get samplesSeen => samplesSeenLo + samplesSeenHi * 0x100000000;
  int get publishNs => publishNsLo + publishNsHi * 0x100000000;
}

/// Header snapshot — same reuse discipline.
class WeftHeaderSnapshot {
  int publishSeqLo = 0, publishSeqHi = 0;
  int epochLo = 0, epochHi = 0;
  int lastPublishNsLo = 0, lastPublishNsHi = 0;
  int framesDroppedLo = 0, framesDroppedHi = 0;
  double globalMin = 0, globalMax = 0, globalAvg = 0, globalCurrent = 0;
  int tickHz = 0, flags = 0;
}

class WeftHotPlane {
  final Uint8List bytes;
  final ByteData _dv;
  final Hpl1Geometry geo;
  final int maxTries;
  int tears = 0;
  int _lastEpochLo = -1, _lastEpochHi = -1;
  late final Uint32List _lastMask;

  WeftHotPlane._(this.bytes, {int byteOffset = 0, this.maxTries = 64})
      : _dv = ByteData.sublistView(bytes, byteOffset),
        geo = Hpl1.validate(bytes, byteOffset: byteOffset) {
    _lastMask = Uint32List(geo.dirtyWords);
  }

  factory WeftHotPlane.fromBytes(Uint8List bytes, {int byteOffset = 0, int maxTries = 64}) {
    return WeftHotPlane._(bytes, byteOffset: byteOffset, maxTries: maxTries);
  }

  /// dart:ffi attach: native plane memory becomes a Dart view. NO COPY.
  /// `ptr` must be 8-byte aligned (HPL1 planes are, by construction).
  factory WeftHotPlane.fromPointer(int address, int byteLength,
      {int maxTries = 64}) {
    // Callers on native platforms pass a Pointer<Uint8>.asTypedList result —
    // kept as a plain factory-over-bytes here so this file stays a pure
    // Dart dependency-free core (the FFI shim lives in weft_hot_plane_notifier
    // example wiring). See example/lib/main.dart for the posix_shm attach.
    throw UnsupportedError(
        'Use WeftHotPlane.fromBytes over a Pointer.asTypedList() view — '
        'see example/lib/main.dart');
  }

  int get laneCount => geo.laneCount;
  int get samplesPerLane => geo.samplesPerLane;

  int _loadSeqLo(int byteOff) => _dv.getUint32(byteOff, Endian.little);

  /// Header seqlock acquire. Returns [Hpl1.ok], [Hpl1.epochChanged] (out still
  /// filled with the NEW header) or [Hpl1.tornSeqlock] (out untouched).
  int readHeader(WeftHeaderSnapshot out) {
    for (var t = 0; t < maxTries; t++) {
      final s1 = _loadSeqLo(Hpl1.hdrPublishSeq);
      if ((s1 & 1) != 0) {
        tears++;
        continue;
      }
      out.publishSeqLo = s1;
      out.publishSeqHi = _dv.getUint32(Hpl1.hdrPublishSeq + 4, Endian.little);
      out.epochLo = _dv.getUint32(Hpl1.hdrEpoch, Endian.little);
      out.epochHi = _dv.getUint32(Hpl1.hdrEpoch + 4, Endian.little);
      out.lastPublishNsLo = _dv.getUint32(Hpl1.hdrLastPublishNs, Endian.little);
      out.globalMin = _dv.getFloat64(Hpl1.hdrGlobalMin, Endian.little);
      out.globalMax = _dv.getFloat64(Hpl1.hdrGlobalMax, Endian.little);
      out.globalAvg = _dv.getFloat64(Hpl1.hdrGlobalAvg, Endian.little);
      out.globalCurrent = _dv.getFloat64(Hpl1.hdrGlobalCurrent, Endian.little);
      out.tickHz = _dv.getUint32(Hpl1.hdrTickHz, Endian.little);
      out.flags = _dv.getUint32(Hpl1.hdrFlags, Endian.little);
      final s2 = _loadSeqLo(Hpl1.hdrPublishSeq);
      if (s1 != s2) {
        tears++;
        continue;
      }
      if (out.epochLo != _lastEpochLo || out.epochHi != _lastEpochHi) {
        final first = _lastEpochLo == -1 && _lastEpochHi == -1;
        _lastEpochLo = out.epochLo;
        _lastEpochHi = out.epochHi;
        if (!first) return Hpl1.epochChanged;
      }
      return Hpl1.ok;
    }
    return Hpl1.tornSeqlock;
  }

  /// Lane seqlock acquire → fills [out] in place. Returns [Hpl1.ok] or
  /// [Hpl1.tornSeqlock] (out untouched — value NOT returned, HPL1 §4.2).
  /// Throws [Hpl1Exception] for caller bugs (lane out of range).
  int readLane(int lane, WeftLaneSnapshot out) {
    if (lane < 0 || lane >= geo.laneCount) {
      throw Hpl1Exception(Hpl1.laneOutOfRange, 'lane $lane of ${geo.laneCount}');
    }
    final ctrl = geo.laneCtrl(lane);
    for (var t = 0; t < maxTries; t++) {
      final s1 = _loadSeqLo(ctrl + Hpl1.laneSeq);
      if ((s1 & 1) != 0) {
        tears++;
        continue;
      }
      out.seqLo = s1;
      out.seqHi = _dv.getUint32(ctrl + Hpl1.laneSeq + 4, Endian.little);
      out.current = _dv.getFloat64(ctrl + Hpl1.laneCurrent, Endian.little);
      out.min = _dv.getFloat64(ctrl + Hpl1.laneMin, Endian.little);
      out.max = _dv.getFloat64(ctrl + Hpl1.laneMax, Endian.little);
      out.avg = _dv.getFloat64(ctrl + Hpl1.laneAvg, Endian.little);
      out.samplesSeenLo = _dv.getUint32(ctrl + Hpl1.laneSamplesSeen, Endian.little);
      out.samplesSeenHi = _dv.getUint32(ctrl + Hpl1.laneSamplesSeen + 4, Endian.little);
      out.head = _dv.getUint32(ctrl + Hpl1.laneHead, Endian.little);
      out.flags = _dv.getUint32(ctrl + Hpl1.laneFlags, Endian.little);
      out.publishNsLo = _dv.getUint32(ctrl + Hpl1.lanePublishNs, Endian.little);
      out.drops = _dv.getUint32(ctrl + Hpl1.laneDrops, Endian.little);
      final s2 = _loadSeqLo(ctrl + Hpl1.laneSeq);
      if (s1 != s2) {
        tears++;
        continue;
      }
      return Hpl1.ok;
    }
    return Hpl1.tornSeqlock;
  }

  /// Copy the k NEWEST samples of [lane] into [out] (index 0 = newest).
  /// Returns samples written (< k ⇒ underrun), or -1 when torn.
  int readRecent(int lane, int k, Float64List out) {
    if (lane < 0 || lane >= geo.laneCount) {
      throw Hpl1Exception(Hpl1.laneOutOfRange, 'lane $lane of ${geo.laneCount}');
    }
    final ctrl = geo.laneCtrl(lane);
    final ringByte = geo.laneRing(lane);
    final mask = geo.samplesPerLane - 1;
    for (var t = 0; t < maxTries; t++) {
      final s1 = _loadSeqLo(ctrl + Hpl1.laneSeq);
      if ((s1 & 1) != 0) {
        tears++;
        continue;
      }
      final head = _dv.getUint32(ctrl + Hpl1.laneHead, Endian.little);
      final seen = _dv.getUint32(ctrl + Hpl1.laneSamplesSeen, Endian.little);
      final s2 = _loadSeqLo(ctrl + Hpl1.laneSeq);
      if (s1 != s2) {
        tears++;
        continue;
      }
      final avail = seen < geo.samplesPerLane ? seen : geo.samplesPerLane;
      final n = k < avail ? k : avail;
      for (var j = 0; j < n; j++) {
        out[j] = _dv.getFloat64(ringByte + ((head - 1 - j) & mask) * 8, Endian.little);
      }
      return n;
    }
    return -1;
  }

  /// Dirty-mask transition scan. Fills [changedOut] with newly-dirty lanes;
  /// returns how many were written. Consumers NEVER write the mask (HPL1 §3).
  int scanDirty(Uint32List changedOut) {
    var written = 0;
    var count = 0;
    for (var w = 0; w < geo.dirtyWords; w++) {
      final cur = _dv.getUint32(Hpl1.headerSize + w * 4, Endian.little);
      final fresh = cur & ~_lastMask[w];
      _lastMask[w] = cur;
      if (fresh != 0 && changedOut != null) {
        var bits = fresh;
        while (bits != 0) {
          final bit = bits & -bits;
          final lane = (w << 5) + bit.bitLength - 1;
          if (written < changedOut.length) changedOut[written++] = lane;
          bits ^= bit;
        }
      }
      var c = cur;
      while (c != 0) {
        c &= c - 1;
        count++;
      }
    }
    _dirtyCount = count;
    return written;
  }

  int _dirtyCount = 0;
  int get dirtyCount => _dirtyCount;
}

File: packages/swift-heddle/Sources/WeftSwiftUI/HotPlane.swift
// HotPlane.swift — HPL1 constants, geometry, and the seqlock consumer over
// POSIX shared memory (or any mapped region). Byte-parity with
// packages/heddle-core/src/layout.js (audited by Tests/ConstantsParity + the
// Node structural audit in audit/static_audit.mjs).
//
// Laws:
//   1  zero allocation on the hot path — readLane fills caller-owned
//      snapshots; no arrays/objects are created per frame
//   2  every multi-byte load is EXPLICIT little-endian
//      (UInt32/UInt64/Double(littleEndian:))
//   3  managed-side only — core/c/weft.{c,h} untouched
//   4  tears/epoch changes are coded outcomes, never silent

import Foundation

/// HPL1 error taxonomy (stable numeric codes — mirrors errors.js).
public enum HPL1Code: Int32, CaseIterable, Sendable {
    case ok = 0
    case badMagic = 1
    case badVersion = 2
    case notLittleEndian = 3
    case capacityMismatch = 4
    case laneOutOfRange = 5
    case tornSeqlock = 6
    case epochChanged = 7
    case planeDetached = 8
    case contextLost = 9
    case tabHidden = 10
    case workerCrash = 11
    case badRenderEngine = 12
    case invalidSample = 13
    case ringUnderrun = 14
}

public struct HPL1 {
    public static let headerSize = 128
    public static let laneCtrlStride = 64
    public static let magicU32: UInt32 = 0x314C_5048 // 'H','P','L','1' LE
    public static let version: UInt32 = 1
    public static let flagLeRequired: UInt32 = 1
    public static let laneFlagActive: UInt32 = 1
    public static let laneFlagManual: UInt32 = 2
    public static let maxLanes = 4096

    // header offsets
    public static let hdrMagic = 0x00
    public static let hdrVersion = 0x04
    public static let hdrFlags = 0x08
    public static let hdrLaneCount = 0x0C
    public static let hdrSamplesPerLane = 0x10
    public static let hdrRingMask = 0x14
    public static let hdrTickHz = 0x18
    public static let hdrReserved0 = 0x1C
    public static let hdrEpoch = 0x20
    public static let hdrPublishSeq = 0x28
    public static let hdrLastPublishNs = 0x30
    public static let hdrFramesDropped = 0x38
    public static let hdrGlobalMin = 0x40
    public static let hdrGlobalMax = 0x48
    public static let hdrGlobalAvg = 0x50
    public static let hdrGlobalCurrent = 0x58
    public static let hdrDirtyWords = 0x60
    public static let hdrLaneCtrlStride = 0x64
    public static let hdrRingBase = 0x68
    public static let hdrTotalBytes = 0x6C

    // lane control block offsets
    public static let laneSeq = 0x00
    public static let laneCurrent = 0x08
    public static let laneMin = 0x10
    public static let laneMax = 0x18
    public static let laneAvg = 0x20
    public static let laneSamplesSeen = 0x28
    public static let laneHead = 0x30
    public static let laneFlags = 0x34
    public static let lanePublishNs = 0x38
    public static let laneDrops = 0x40

    static func align8(_ x: Int) -> Int { (x + 7) & ~7 }
    static func isPow2(_ x: Int) -> Bool { x >= 2 && (x & (x - 1)) == 0 }

    public static func deriveGeometry(laneCount: Int, samplesPerLane: Int) -> HPL1Geometry {
        let dirtyWords = (laneCount + 31) / 32
        let laneCtrlBase = headerSize + align8(4 * dirtyWords)
        let ringBase = align8(laneCtrlBase + laneCtrlStride * laneCount)
        let totalBytes = ringBase + laneCount * samplesPerLane * 8
        return HPL1Geometry(dirtyWords: dirtyWords, laneCtrlBase: laneCtrlBase,
                            ringBase: ringBase, totalBytes: totalBytes,
                            laneCount: laneCount, samplesPerLane: samplesPerLane)
    }

    static func validate(_ bytes: UnsafeRawBufferPointer, byteOffset: Int = 0) throws -> HPL1Geometry {
        let available = bytes.count - byteOffset
        guard available >= headerSize else { throw HPL1Error(.planeDetached, "buffer too small") }
        let base = byteOffset
        func u32(_ off: Int) -> UInt32 { bytes.loadUnaligned(fromByteOffset: base + off, as: UInt32.self).littleEndian }
        guard u32(hdrMagic) == magicU32 else { throw HPL1Error(.badMagic, "magic mismatch") }
        guard u32(hdrVersion) == version else { throw HPL1Error(.badVersion, "version mismatch") }
        guard u32(hdrFlags) & flagLeRequired != 0 else { throw HPL1Error(.notLittleEndian, "LE_REQUIRED clear") }
        let laneCount = Int(u32(hdrLaneCount))
        let samplesPerLane = Int(u32(hdrSamplesPerLane))
        guard (1...maxLanes).contains(laneCount) else { throw HPL1Error(.capacityMismatch, "laneCount range") }
        guard isPow2(samplesPerLane) else { throw HPL1Error(.capacityMismatch, "samplesPerLane not pow2") }
        guard u32(hdrRingMask) == UInt32(samplesPerLane - 1) else { throw HPL1Error(.capacityMismatch, "ringMask drift") }
        let geo = deriveGeometry(laneCount: laneCount, samplesPerLane: samplesPerLane)
        guard u32(hdrDirtyWords) == UInt32(geo.dirtyWords),
              u32(hdrLaneCtrlStride) == UInt32(laneCtrlStride),
              u32(hdrRingBase) == UInt32(geo.ringBase),
              u32(hdrTotalBytes) == UInt32(geo.totalBytes) else {
            throw HPL1Error(.capacityMismatch, "derived vs stored drift")
        }
        guard available >= geo.totalBytes else { throw HPL1Error(.planeDetached, "buffer smaller than plane") }
        return geo
    }
}

public struct HPL1Geometry: Sendable {
    public let dirtyWords: Int
    public let laneCtrlBase: Int
    public let ringBase: Int
    public let totalBytes: Int
    public let laneCount: Int
    public let samplesPerLane: Int
    public func laneCtrl(_ lane: Int) -> Int { laneCtrlBase + lane * HPL1.laneCtrlStride }
    public func laneRing(_ lane: Int) -> Int { ringBase + lane * samplesPerLane * 8 }
}

public struct HPL1Error: Error, Sendable {
    public let code: HPL1Code
    public let detail: String
    public init(_ code: HPL1Code, _ detail: String) { self.code = code; self.detail = detail }
}

/// Caller-owned lane snapshot (allocate once, reuse — Law 1).
public struct WeftLaneSnapshot {
    public var seqLo: UInt32 = 0, seqHi: UInt32 = 0
    public var current: Double = 0, min: Double = 0, max: Double = 0, avg: Double = 0
    public var samplesSeenLo: UInt32 = 0, samplesSeenHi: UInt32 = 0
    public var head: UInt32 = 0, flags: UInt32 = 0
    public var publishNsLo: UInt32 = 0, publishNsHi: UInt32 = 0
    public var drops: UInt32 = 0
    public var samplesSeen: UInt64 { UInt64(samplesSeenLo) | (UInt64(samplesSeenHi) << 32) }
}

public struct WeftHeaderSnapshot {
    public var publishSeqLo: UInt32 = 0, publishSeqHi: UInt32 = 0
    public var epochLo: UInt32 = 0, epochHi: UInt32 = 0
    public var lastPublishNsLo: UInt32 = 0, lastPublishNsHi: UInt32 = 0
    public var globalMin: Double = 0, globalMax: Double = 0, globalAvg: Double = 0, globalCurrent: Double = 0
    public var tickHz: UInt32 = 0, flags: UInt32 = 0
}

/// WeftHotPlane — seqlock consumer over mapped memory. NOT Sendable: each
/// thread owns its instance (tears/epoch caches are per-reader by design).
public final class WeftHotPlane {
    public let bytes: UnsafeRawBufferPointer
    public let ownsMemory: Bool
    public let geo: HPL1Geometry
    public let maxTries: Int
    public internal(set) var tears: Int = 0
    var lastEpochLo: UInt32 = .max // .max ≡ "primed" sentinel (0xFFFFFFFF never a real epoch-lo start)
    var lastEpochHi: UInt32 = .max
    var lastMask: [UInt32]
    public internal(set) var dirtyCount: Int = 0

    /// Attach an existing mapped region (Engineer 1's bridge / POSIX shm).
    /// The buffer is NOT copied and NOT freed by this class (ownsMemory=false).
    public init(bytes: UnsafeRawBufferPointer, byteOffset: Int = 0, maxTries: Int = 64) throws {
        self.bytes = UnsafeRawBufferPointer(rebasing: bytes[byteOffset...])
        self.ownsMemory = false
        self.maxTries = maxTries
        self.geo = try HPL1.validate(self.bytes)
        self.lastMask = Array(repeating: 0, count: geo.dirtyWords)
    }

    deinit {
        // ownership of externally-provided memory stays with the caller
    }

    var laneCount: Int { geo.laneCount }
    var samplesPerLane: Int { geo.samplesPerLane }

    func loadSeqLo(_ byteOff: Int) -> UInt32 {
        bytes.loadUnaligned(fromByteOffset: byteOff, as: UInt32.self).littleEndian
    }

    /// Header seqlock acquire → fills `out` in place. Returns .ok,
    /// .epochChanged (out still filled) or .tornSeqlock (out untouched).
    @discardableResult
    public func readHeader(_ out: inout WeftHeaderSnapshot) -> HPL1Code {
        for _ in 0..<maxTries {
            let s1 = loadSeqLo(HPL1.hdrPublishSeq)
            if s1 & 1 != 0 { tears += 1; continue }
            out.publishSeqLo = s1
            out.publishSeqHi = bytes.loadUnaligned(fromByteOffset: HPL1.hdrPublishSeq + 4, as: UInt32.self).littleEndian
            out.epochLo = bytes.loadUnaligned(fromByteOffset: HPL1.hdrEpoch, as: UInt32.self).littleEndian
            out.epochHi = bytes.loadUnaligned(fromByteOffset: HPL1.hdrEpoch + 4, as: UInt32.self).littleEndian
            out.lastPublishNsLo = bytes.loadUnaligned(fromByteOffset: HPL1.hdrLastPublishNs, as: UInt32.self).littleEndian
            out.globalMin = bytes.loadUnaligned(fromByteOffset: HPL1.hdrGlobalMin, as: UInt64.self).littleEndian.bitPattern
            out.globalMax = bytes.loadUnaligned(fromByteOffset: HPL1.hdrGlobalMax, as: UInt64.self).littleEndian.bitPattern
            out.globalAvg = bytes.loadUnaligned(fromByteOffset: HPL1.hdrGlobalAvg, as: UInt64.self).littleEndian.bitPattern
            out.globalCurrent = bytes.loadUnaligned(fromByteOffset: HPL1.hdrGlobalCurrent, as: UInt64.self).littleEndian.bitPattern
            out.tickHz = bytes.loadUnaligned(fromByteOffset: HPL1.hdrTickHz, as: UInt32.self).littleEndian
            out.flags = bytes.loadUnaligned(fromByteOffset: HPL1.hdrFlags, as: UInt32.self).littleEndian
            let s2 = loadSeqLo(HPL1.hdrPublishSeq)
            if s1 != s2 { tears += 1; continue }
            if out.epochLo != lastEpochLo || out.epochHi != lastEpochHi {
                let first = lastEpochLo == .max && lastEpochHi == .max
                lastEpochLo = out.epochLo
                lastEpochHi = out.epochHi
                if !first { return .epochChanged }
            }
            return .ok
        }
        return .tornSeqlock
    }

    /// Lane seqlock acquire → fills `out` in place. .ok | .tornSeqlock.
    @discardableResult
    public func readLane(_ lane: Int, _ out: inout WeftLaneSnapshot) -> HPL1Code {
        guard lane >= 0 && lane < geo.laneCount else {
            return .laneOutOfRange // caller error surfaced as a typed code
        }
        let ctrl = geo.laneCtrl(lane)
        for _ in 0..<maxTries {
            let s1 = loadSeqLo(ctrl + HPL1.laneSeq)
            if s1 & 1 != 0 { tears += 1; continue }
            out.seqLo = s1
            out.seqHi = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneSeq + 4, as: UInt32.self).littleEndian
            out.current = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneCurrent, as: UInt64.self).littleEndian.bitPattern
            out.min = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneMin, as: UInt64.self).littleEndian.bitPattern
            out.max = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneMax, as: UInt64.self).littleEndian.bitPattern
            out.avg = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneAvg, as: UInt64.self).littleEndian.bitPattern
            out.samplesSeenLo = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneSamplesSeen, as: UInt32.self).littleEndian
            out.samplesSeenHi = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneSamplesSeen + 4, as: UInt32.self).littleEndian
            out.head = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneHead, as: UInt32.self).littleEndian
            out.flags = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneFlags, as: UInt32.self).littleEndian
            out.publishNsLo = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.lanePublishNs, as: UInt32.self).littleEndian
            out.drops = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneDrops, as: UInt32.self).littleEndian
            let s2 = loadSeqLo(ctrl + HPL1.laneSeq)
            if s1 != s2 { tears += 1; continue }
            return .ok
        }
        return .tornSeqlock
    }

    /// Copy the k NEWEST samples into `out` (index 0 = newest).
    /// Returns count written (< k ⇒ underrun), or -1 when torn.
    public func readRecent(_ lane: Int, _ k: Int, _ out: UnsafeMutableBufferPointer<Double>) -> Int {
        guard lane >= 0 && lane < geo.laneCount else { return -2 }
        let ctrl = geo.laneCtrl(lane)
        let ringByte = geo.laneRing(lane)
        let mask = geo.samplesPerLane - 1
        for _ in 0..<maxTries {
            let s1 = loadSeqLo(ctrl + HPL1.laneSeq)
            if s1 & 1 != 0 { tears += 1; continue }
            let head = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneHead, as: UInt32.self).littleEndian
            let seen = bytes.loadUnaligned(fromByteOffset: ctrl + HPL1.laneSamplesSeen, as: UInt32.self).littleEndian
            let s2 = loadSeqLo(ctrl + HPL1.laneSeq)
            if s1 != s2 { tears += 1; continue }
            let avail = Int(min(UInt(seen), UInt(geo.samplesPerLane)))
            let n = min(k, avail)
            for j in 0..<n {
                // NOTE: Swift's `&` binds LOOSER than `*` — mask first, then scale
                let slot = ((Int(head) - 1 - j) & mask) * 8
                let bits = bytes.loadUnaligned(fromByteOffset: ringByte + slot, as: UInt64.self).littleEndian
                out[j] = bits.bitPattern
            }
            return n
        }
        return -1
    }

    /// Dirty-mask transition scan; newly-dirty lanes returned (HPL1 §3 —
    /// consumers NEVER write the mask).
    public func scanDirty(_ changedOut: inout [Int]) -> Int {
        changedOut.removeAll(keepingCapacity: true)
        var count = 0
        for w in 0..<geo.dirtyWords {
            let cur = bytes.loadUnaligned(fromByteOffset: HPL1.headerSize + w * 4, as: UInt32.self).littleEndian
            let fresh = cur & ~lastMask[w]
            lastMask[w] = cur
            if fresh != 0 {
                var bits = fresh
                while bits != 0 {
                    let bit = bits & (~bits &+ 1)
                    let lane = w * 32 + bit.trailingZeroBitCount
                    changedOut.append(lane)
                    bits &= bits &- 1
                }
            }
            count += cur.nonzeroBitCount
        }
        dirtyCount = count
        return changedOut.count
    }
}
